align_system/algorithms/outlines_adm.py

```diff
@@ -150,18 +255,81 @@ def top_level_choose_action(self,
+                    possible_icl_examples.append({
+                        "state": state,
+                        "actions": actions,
+                        "labels": sample["label"]
+                    })
+
+                # Downselect to n_icl_examples via given method
+                icl_strategy = kwargs["incontext"]["method"]
+                if icl_strategy == "random":
+                    selected_icl_examples = random.sample(possible_icl_examples, n_icl_examples)
+                elif icl_strategy == "bert_similarity":
+                    # Only comparing similarity of prompts, not considering answer options at this time
+                    no_choices_prompt, _ = self._state_to_top_level_prompt(scenario_state, [])
+                    possible_icl_prompts = []
+                    for s in possible_icl_examples:
+                        prompt, _ = self._state_to_top_level_prompt(s["state"], [])
+                        possible_icl_prompts.append(prompt)
+
+                    # Create similarity scores between the ICL samples and find top-k indices
+                    from bert_score import score
+                    _, _, F1 = score([no_choices_prompt]*len(possible_icl_prompts), possible_icl_prompts, lang="en")
+                    _, indices = torch.topk(F1, n_icl_examples)
+
+                    selected_icl_examples = [possible_icl_examples[i] for i in indices]
+                else:
+                    raise ValueError(f'"{icl_strategy}" is not a valid incontext method. Please use "random" or '
+                                      '"bert_similarity"')
+
+                # Create ICL prompts
+                for sample in selected_icl_examples:
+                    positive_icl_examples.extend(
+                        self._format_single_incontext_prompt(sample, kdma, value)
+                    )
+
+                    if num_negative_samples > 0:
+                        negative_icl_examples.extend(
+                            self._format_single_incontext_prompt(sample, kdma, negative_value)
+                        )
         else:
             positive_system_prompt = baseline_system_prompt()
             if num_negative_samples > 0:
                 raise RuntimeError("No notion of negative samples for baseline run")
+            if "incontext" in kwargs and kwargs["incontext"]["number"] > 0:
+                raise RuntimeError("No notion of incontext examples for baseline run")
 
         positive_dialogs = []
         for _ in range(num_positive_samples):
             shuffled_choices = random.sample(choices, len(choices))
 
             prompt = action_selection_prompt(scenario_description, shuffled_choices)
-            dialog = [{'role': 'system', 'content': positive_system_prompt},
-                      {'role': 'user', 'content': prompt}]
+            dialog = [{'role': 'system', 'content': positive_system_prompt}]
+            dialog.extend(positive_icl_examples)
```

Comment:
I would really like this call here to mostly be the extent of the reference / call to in-context examples.

What I mean is that I think all of the in context state hydration, searching, and preparation can live in it's own file (and class is probably ideal so that it can be configured / instantiated from Hydra) since there are very few dependencies on what's happening in this file (if that makes sense).  Would like to see a function on the in-context class that's something like `retrieve_incontext_examples` where you give it the number of positive and negative examples you want etc.